### PR TITLE
[Enhancement][FSDP] Only recognize module `mixed gradients` as `ignored_modules`

### PR DESCRIPTION
**Pull request by HAOCHENYE**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

There is an issue with PyTorch's FSDP (Fully Sharded DataParallel) where it cannot directly wrap models that have some parameters with gradients and some parameters without gradients. FSDP requires that all parameters of the model have the same gradient attribute, meaning either all parameters have gradients or none of them do. 

Therefore, we need to inform FSDP about a small subset of special parameters (either with gradients or without gradients) by passing them as ignored_modules. This way, these modules will not be sharded and will not cause any errors.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmengine/model/wrappers/fully_sharded_distributed.py [hunk 1]
```diff
@@ -219,9 +219,10 @@ def parse_dtype(dtype):
                 '`mixed_precision` should be `None`, `dict` or '
                 f'`MixedPrecision`, but has type {type(mixed_precision)}')
 
-        self._fixed_modules = self._get_fixed_module(module, ignored_modules)
         ignored_modules = [] if ignored_modules is None else ignored_modules
-        ignored_modules = chain(ignored_modules, self._fixed_modules)
+        self._auto_ignored_modules = self._find_ignored_module(
```
Comment (zhouzaida) [inline mmengine/model/wrappers/fully_sharded_distributed.py:223]: The attribute is not necessary.

Comment (zhouzaida) [reply mmengine/model/wrappers/fully_sharded_distributed.py:223]: A normal variable is enough.

File: mmengine/model/wrappers/fully_sharded_distributed.py [hunk 2]
```diff
@@ -323,30 +332,66 @@ def _get_fixed_module(self, module, ignored_modules):
...
+        #    recognized as an ignored module if its parameters is `mixed`
+        # 2. `module._parameters` is empty, such as `ResNet50`, it will be
+        #    recognized as an ignored module if both `true` module and `false`
+        #    module exist in its submodules.
+        # 3. For other cases, if its parameters are `mixed` or the state of
+        #    its submodules diff from the state of the parameters, it will be
+        #    recognized an ignored module.
+        def find_ignored_modules(root_module: nn.Module):
+            if not list(root_module.children()):
+                if 'mixed' == get_grad_state(root_module):
```
Comment (zhouzaida) [inline mmengine/model/wrappers/fully_sharded_distributed.py:361]:
```suggestion
                if get_grad_state(root_module) == 'mixed':
```

File: mmengine/model/wrappers/fully_sharded_distributed.py [hunk 3]
```diff
@@ -323,30 +332,66 @@ def _get_fixed_module(self, module, ignored_modules):
...
+                children = list(root_module.children())
+                params = list(root_module.parameters(recurse=False))
+                if children and not params:
+                    states = list(get_grad_state(m) for m in children)
+                    if not (all(s == 'true' for s in states)
+                            or all(s == 'false'
+                                   for s in states) or 'mixed' in states):
+                        auto_ignored_modules.append(root_module)
+                        return
+                elif params and children:
```
Comment (zhouzaida) [inline mmengine/model/wrappers/fully_sharded_distributed.py:373]:
```suggestion
                elif children and module:
```

File: mmengine/model/wrappers/fully_sharded_distributed.py [hunk 4]
```diff
@@ -323,30 +332,66 @@ def _get_fixed_module(self, module, ignored_modules):
...
+        # Here are some rules to recognize the ignored modules:
+        # 1. If `module._modules` is empty, such as `nn.Linear`, it will be
+        #    recognized as an ignored module if its parameters is `mixed`
+        # 2. `module._parameters` is empty, such as `ResNet50`, it will be
+        #    recognized as an ignored module if both `true` module and `false`
+        #    module exist in its submodules.
+        # 3. For other cases, if its parameters are `mixed` or the state of
+        #    its submodules diff from the state of the parameters, it will be
+        #    recognized an ignored module.
+        def find_ignored_modules(root_module: nn.Module):
```
Comment (zhouzaida) [inline mmengine/model/wrappers/fully_sharded_distributed.py:359]:
```suggestion
        def find_mixed_modules(root_module: nn.Module):
```
